hw/ip/otbn/util/shared/control_flow.py

```diff
@@ -0,0 +1,392 @@
+                format(pc, ['{:#x}'.format(pc) for pc in self.edges.keys()]))
+        return self.edges[pc]
+
+    def get_cycles(self, pc, path=[]) -> Dict[int, List[int]]:
+        '''Returns any control-flow cycles in this graph.'''
+        if pc not in self.sections:
+            raise ValueError(
+                'Invalid entry PC when getting cycles: {:#x}. Valid '
+                'control PCs are: {}'.format(
+                    pc, ['{:#x}'.format(pc) for pc in self.sections.keys()]))
+        cycles: Dict[int, List[int]] = {}
+        sec = self.get_section(pc)
+        for loc in self.get_edges(pc):
+            if isinstance(loc, LoopStart):
+                # If this is the last loop instruction, don't loop back
+                if sec.end == loc.loop_end_pc:
+                    continue
+                pc = loc.loop_start_pc
+            elif loc.is_special():
+                # Skip Ret, Ecall, ImemEnd
+                continue
+            else:
+                pc = loc.pc
+            if pc in path:
+                # Found a cycle; report it
+                if pc not in cycles:
+                    cycles[pc] = []
+                cycles[pc].append(sec.end)
+            else:
+                cycles.update(
+                    self.get_cycles(
+                        pc, path + list(range(sec.start, sec.end + 4, 4))))
+        return cycles
+
+    def _pretty_lines(self,
+                      program,
+                      entry_pc,
+                      concise,
+                      already_printed,
+                      call_stack=[]):
+        '''Returns lines for pretty-printing.
+
+        Returns each line as an (int,str) pair, where the int is the indent
+        level.
+        '''
+        if entry_pc not in self.sections:
+            raise ValueError(
+                'Invalid entry PC when pretty-printing control graph: '
+                '{:#x}. Valid control PCs are: {}'.format(
+                    entry_pc,
+                    ['{:#x}'.format(pc) for pc in self.sections.keys()]))
+        out = []
+        sec = self.sections[entry_pc]
+        symbols = program.get_symbols_for_pc(entry_pc)
+        pcs_to_print = range(sec.start, sec.end + 4, 4)
+        if len(symbols) > 0:
+            out.append((0, '<{}>'.format(', '.join(symbols))))
+            if entry_pc in already_printed and not concise:
+                out.append((0, '(see above)'))
+                pcs_to_print = []
+        if concise:
+            # only print last insn
+            out.append((0, '...'))
+            pcs_to_print = [sec.end]
+        for pc in pcs_to_print:
+            insn = program.get_insn(pc)
+            operands = program.get_operands(pc)
+            out.append((0, '0x{:x}: {}'.format(pc,
+                                               insn.disassemble(pc,
+                                                                operands))))
```

Comment:
Oof, yeah, that's not so nice. My usual workflow is to write the code and then `lintpy --fix` it; do you have another recommendation?

Reply:
I tend not to use `lintpy --fix`: I don't *think* it's mandated by the project, and I think I format things better :-) But if you're using an auto-formatter like that, maybe the trick is to check things in before running it so that you can look at the diffs for anything suspect?